gen3-cli/cot/backend/common/context.py

```diff
@@ -0,0 +1,233 @@
+
+    levels = attrdict()
+
+    @property
+    def dir(self):
+        return self.__dir
+
+    @dir.setter
+    def dir(self, value):
+        value = os.path.normcase(value)
+        if not os.path.isabs(value):
+            raise ValueError('{} is not abs path'.format(value))
+        if not os.path.isdir(value):
+            raise ValueError('{} is not dir'.format(value))
+        self.__dir = value
+        self.__search = ContextSearch(self.__dir)
+        self.__try_to_set_root()
+        self.__try_to_set_tenant()
+
+    @property
+    def root(self):
+        return self.__root
+
+    @property
+    def tenant(self):
+        return self.__tenant
+
+    @property
+    def account(self):
+        return self.__account
+
+    @property
+    def search(self):
+        return self.__search
+
+    def __init__(self, dir, account=None):
+        self.dir = dir
+        self.props = ContextProps()
+        self.__account = account
+        if hasattr(self, 'filename'):
+            self.__check_level_filename()
+            self.setup()
+
+    def __check_level_filename(self):
+        if not self.search.isfile(self.filename):
+            raise NoLevelFileError(self.filename)
+
+    def __try_to_set_root(self):
+        self.__root = os.path.dirname(self.__search.upwards(self.levels.root.filename))
+        if not self.__root:
+            raise ValueError('Unable to find {} in {} or its parents'.format(self.levels.root.filename, self.__dir))
+
+    def __try_to_set_account(self):
+        found = Search.downwards(self.root, self.levels.account.filename)
+        if not found:
+            raise ContextError('No accounts found')
+        selected_account = None
+        found_accounts = []
+        for account_filename in found:
+            account_level_dir = self.levels.account(os.path.basename(account_filename))
+            try:
+                account = self.levels.account(account_level_dir)
+                found_accounts.append(account)
+                if self.account and account.props.name == self.account:
+                    selected_account = account
+            except ContextError as e:
+                raise ContextError('Corrupted account found in {}'.format(account_level_dir)) from e
+        if self.account is not None:
+            if selected_account is None:
+                raise ContextError("Can't find account {}".format(self.account))
+            else:
+                self.account = account
+        else:
+            if len(found_accounts) > 1:
+                raise ContextError("Can't automatically determine which account to use. Please specify account name.")
+            else:
+                self.account = found_accounts[0]
+
+    def __try_to_set_tenant(self):
+        found = Search.downwards(self.root, self.levels.tenant.filename)
+        if not found:
+            raise ContextError('Tenant not found')
+
+    def __str__(self):
+        template = "level: $name\ndir: $dir\nprops:\n$props"
```

Comment:
I just wanted to make it multiline, also I'm working on python 3.6. I probably need to spend some time upgrading python to 3.7 because python 3.6 is lacking a lot of nice quality of life features.

Reply:
yes, I believe we are not bound to specific python version, so it's fine to use more or less recent software.
I'm not sure if our clients will be running it in the docker exclusively or on their normal client environments; in first case we can use bleeding-edge Python, but in second it's worth even sticking to the 3.4 or something ancient likewise (so popular repositories can run it without docker, using their system default Python version).
hard to say.